File: ios/FaceBlood/FaceBlood/WaveformView.swift
//
//  WaveformView.swift
//  FaceBlood
//
//  Lightweight rPPG waveform plotter.
//

import SwiftUI

struct WaveformView: View {
    let data: [Float]
    let mode: PulseMode
    let active: Bool

    var body: some View {
        Canvas { ctx, size in
            // Grid lines.
            let grid = Path { p in
                for i in 0...4 {
                    let y = CGFloat(i) / 4 * size.height
                    p.move(to: CGPoint(x: 0, y: y))
                    p.addLine(to: CGPoint(x: size.width, y: y))
                }
            }
            ctx.stroke(grid,
                       with: .color(.white.opacity(0.06)),
                       lineWidth: 1)

            guard data.count >= 2 else { return }
            var path = Path()
            let step = size.width / CGFloat(data.count - 1)
            for (i, v) in data.enumerated() {
                let x = CGFloat(i) * step
                let y = size.height / 2 - CGFloat(v) * size.height * 0.42
                if i == 0 { path.move(to: CGPoint(x: x, y: y)) }
                else { path.addLine(to: CGPoint(x: x, y: y)) }
            }

            let leadColor = active ? mode.systoleColor : Color(red: 0.47, green: 0.78, blue: 1)
            let trailColor = active ? mode.diastoleColor.opacity(0) : Color(red: 0.47, green: 0.78, blue: 1).opacity(0)

            ctx.stroke(
                path,
                with: .linearGradient(
                    Gradient(colors: [trailColor, leadColor]),
                    startPoint: CGPoint(x: 0, y: 0),
                    endPoint: CGPoint(x: size.width, y: 0)
                ),
                lineWidth: 1.8
            )
        }
    }
}

File: android/app/src/main/java/com/nwlab/faceblood/ui/components/WaveformView.kt
package com.nwlab.faceblood.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import com.nwlab.faceblood.model.PulseMode
import com.nwlab.faceblood.ui.theme.CyanHud

/**
 * Waveform canvas that mirrors [WaveformView.swift] on iOS.
 *
 * Draws five faint horizontal grid lines and a polyline from the [data] array,
 * coloured with a left-to-right gradient from the mode's systole colour to the
 * diastole colour (or a default cyan when inactive).
 */
@Composable
fun WaveformView(
    data: List<Float>,
    mode: PulseMode,
    active: Boolean,
    modifier: Modifier = Modifier,
) {
    Canvas(modifier = modifier) {
        val w = size.width
        val h = size.height
        val mid = h / 2f

        // Grid lines
        val gridPaint = androidx.compose.ui.graphics.drawscope.DrawScope::drawLine
        for (row in 0..4) {
            val y = h * row / 4f
            drawLine(
                color = Color.White.copy(alpha = 0.08f),
                start = Offset(0f, y),
                end = Offset(w, y),
                strokeWidth = 1f,
            )
        }

        if (data.size < 2) return@Canvas

        // Build path
        val path = Path()
        data.forEachIndexed { i, v ->
            val x = w * i / (data.size - 1).toFloat()
            val y = mid - v * h * 0.42f
            if (i == 0) path.moveTo(x, y) else path.lineTo(x, y)
        }

        val startColor = if (active) mode.systoleColor else CyanHud
        val endColor = if (active) mode.diastoleColor.copy(alpha = 0f) else CyanHud.copy(alpha = 0f)

        drawPath(
            path = path,
            brush = Brush.horizontalGradient(listOf(startColor, endColor)),
            style = Stroke(width = 2f, cap = StrokeCap.Round, join = StrokeJoin.Round),
        )
    }
}
